plaso/engine/artifact_filters.py

```diff
@@ -75,71 +74,111 @@ def BuildFindSpecs(self, environment_variables=None):
-
-        elif (source.type_indicator ==
-              artifact_types.TYPE_INDICATOR_ARTIFACT_GROUP):
-          self._artifacts.remove(name)
-          for name_entry in set(source.names):
-            self._artifacts.append(name_entry)
-          self.BuildFindSpecs(environment_variables=environment_variables)
-
-        else:
-          logger.warning(
-              'Unsupported artifact definition source type: "{0:s}"'.format(
-                  source.type_indicator))
+      find_specs.extend(self._BuildFindSpecsFromArtifact(
+          definition, environment_variables))
+
+    find_specs_per_source_type = defaultdict(list)
+    for find_spec in find_specs:
+      if isinstance(find_spec, registry_searcher.FindSpec):
+        artifact_list = find_specs_per_source_type[
+            artifact_types.TYPE_INDICATOR_WINDOWS_REGISTRY_KEY]
+        artifact_list.append(find_spec)
+        continue
+
+      if isinstance(find_spec, file_system_searcher.FindSpec):
+        artifact_list = find_specs_per_source_type[
+            artifact_types.TYPE_INDICATOR_FILE]
+        artifact_list.append(find_spec)
+        continue
+      logger.warning('Unknown find specification type: {0:s}'.format(
+          type(find_spec)))
 
     self._knowledge_base.SetValue(
-        self.KNOWLEDGE_BASE_VALUE, self._find_specs_per_source_type)
+        self.KNOWLEDGE_BASE_VALUE, find_specs_per_source_type)
+
+
+  def _BuildFindSpecsFromArtifact(self, definition, environment_variables):
+    """Builds find specifications from an artifact definition.
+
+    Args:
+      definition (artifacts.ArtifactDefinition): artifact definition.
+      environment_variables (list[EnvironmentVariableArtifact]):
+          environment variables.
+
+    Returns:
+      list[dfwinreg.FindSpec|dfvfs.FindSpec]: find specifications.
+    """
+    find_specs = []
+    for source in definition.sources:
+      if source.type_indicator == artifact_types.TYPE_INDICATOR_FILE:
+        for path_entry in set(source.paths):
+          specifications = self._BuildFindSpecsFromFileSource(
+              path_entry, source.separator, environment_variables,
+              self._knowledge_base.user_accounts)
+          find_specs.extend(specifications)
+
+      elif (source.type_indicator ==
+            artifact_types.TYPE_INDICATOR_WINDOWS_REGISTRY_KEY):
+        for key_path in set(source.keys):
+          if self.CheckKeyCompatibility(key_path):
+            specifications = self._BuildFindSpecsFromRegistrySource(key_path)
+            find_specs.extend(specifications)
+
+      elif (source.type_indicator ==
+            artifact_types.TYPE_INDICATOR_WINDOWS_REGISTRY_VALUE):
+        # TODO: Handle Registry Values Once Supported in dfwinreg.
+        # https://github.com/log2timeline/dfwinreg/issues/98
+        logger.warning((
+            'Windows Registry values are not supported, extracting key: '
+            '"{0!s}"').format(source.key_value_pairs))
```

Comment:
convert `source.key_value_pairs` into human readable text instead of letting this up to `format()`,

Reply:
Done